packages/animations/lib/src/shared_z_axis_transition.dart

```diff
@@ -0,0 +1,394 @@
+    super.initState();
+    _effectiveAnimationStatus = widget.animation.status;
+    _effectiveSecondaryAnimationStatus = widget.secondaryAnimation.status;
+    widget.animation.addStatusListener(_animationListener);
+    widget.secondaryAnimation.addStatusListener(_secondaryAnimationListener);
+  }
+
+  void _animationListener(AnimationStatus animationStatus) {
+    _effectiveAnimationStatus = _calculateEffectiveAnimationStatus(
+      lastEffective: _effectiveAnimationStatus,
+      current: animationStatus,
+    );
+  }
+
+  void _secondaryAnimationListener(AnimationStatus animationStatus) {
+    _effectiveSecondaryAnimationStatus = _calculateEffectiveAnimationStatus(
+      lastEffective: _effectiveSecondaryAnimationStatus,
+      current: animationStatus,
+    );
+  }
+
+  // When a transition is interrupted midway we just want to play the ongoing
+  // animation in reverse. Switching to the actual reverse transition would
+  // yield a disjoint experience since the forward and reverse transitions are
+  // very different.
+  AnimationStatus _calculateEffectiveAnimationStatus({
+    @required AnimationStatus lastEffective,
+    @required AnimationStatus current,
+  }) {
+    assert(current != null);
+    assert(lastEffective != null);
+    switch (current) {
+      case AnimationStatus.dismissed:
+      case AnimationStatus.completed:
+        return current;
+      case AnimationStatus.forward:
+        switch (lastEffective) {
+          case AnimationStatus.dismissed:
+          case AnimationStatus.completed:
+          case AnimationStatus.forward:
+            return current;
+          case AnimationStatus.reverse:
+            return lastEffective;
+        }
+        break;
+      case AnimationStatus.reverse:
+        switch (lastEffective) {
+          case AnimationStatus.dismissed:
+          case AnimationStatus.completed:
+          case AnimationStatus.reverse:
+            return current;
+          case AnimationStatus.forward:
+            return lastEffective;
+        }
+        break;
+    }
+    return null; // unreachable
+  }
+
+  @override
+  void didUpdateWidget(SharedZAxisTransition oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.animation != widget.animation) {
+      oldWidget.animation.removeStatusListener(_animationListener);
+      widget.animation.addStatusListener(_animationListener);
+      _animationListener(widget.animation.status);
+    }
+    if (oldWidget.secondaryAnimation != widget.secondaryAnimation) {
+      oldWidget.secondaryAnimation
+          .removeStatusListener(_secondaryAnimationListener);
+      widget.secondaryAnimation.addStatusListener(_secondaryAnimationListener);
+      _secondaryAnimationListener(widget.secondaryAnimation.status);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.animation.removeStatusListener(_animationListener);
+    widget.secondaryAnimation.removeStatusListener(_secondaryAnimationListener);
+    super.dispose();
```

Comment:
Could this widget have been disposed while the animation (the AnimatedBuilder) is still running?

I think this is safe, since dispose implies that the AnimatedBuilder is disposed as well. It would be good to test this part of the implementation (rebuild a widget with and then without the transition, while the transition is underway).

Reply:
I added a test for this. PTAL